In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/indian-startup-funding/startup_funding.csv


In [3]:
df = pd.read_csv('/kaggle/input/indian-startup-funding/startup_funding.csv')

In [4]:
df.head()

Sr No Date dd/mm/yyyy                  Startup Name    Industry Vertical  \
0      1      09/01/2020                        BYJU’S               E-Tech   
1      2      13/01/2020                        Shuttl       Transportation   
2      3      09/01/2020                     Mamaearth           E-commerce   
3      4      02/01/2020  https://www.wealthbucket.in/              FinTech   
4      5      02/01/2020                        Fashor  Fashion and Apparel   

                             SubVertical City  Location  \
0                             E-learning      Bengaluru   
1              App based shuttle service        Gurgaon   
2  Retailer of baby and toddler products      Bengaluru   
3                      Online Investment      New Delhi   
4            Embroiled Clothes For Women         Mumbai   

              Investors Name       InvestmentnType Amount in USD Remarks  
0    Tiger Global Management  Private Equity Round  20,00,00,000     NaN  
1  Susquehanna Growth Equity              Series C     80,48,394     NaN  
2      Sequoia Capital India              Series B   1,83,58,860     NaN  
3             Vinod Khatumal          Pre-series A     30,00,000     NaN  
4    Sprout Venture Partners            Seed Round     18,00,000     NaN

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3044 entries, 0 to 3043
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Sr No              3044 non-null   int64 
 1   Date dd/mm/yyyy    3044 non-null   object
 2   Startup Name       3044 non-null   object
 3   Industry Vertical  2873 non-null   object
 4   SubVertical        2108 non-null   object
 5   City  Location     2864 non-null   object
 6   Investors Name     3020 non-null   object
 7   InvestmentnType    3040 non-null   object
 8   Amount in USD      2084 non-null   object
 9   Remarks            419 non-null    object
dtypes: int64(1), object(9)
memory usage: 237.9+ KB


#### Removing remarks column 

In [7]:
df.drop(columns=['Remarks'], inplace = True)

In [8]:
df.head(2)

Sr No Date dd/mm/yyyy Startup Name Industry Vertical  \
0      1      09/01/2020       BYJU’S            E-Tech   
1      2      13/01/2020       Shuttl    Transportation   

                 SubVertical City  Location             Investors Name  \
0                 E-learning      Bengaluru    Tiger Global Management   
1  App based shuttle service        Gurgaon  Susquehanna Growth Equity   

        InvestmentnType Amount in USD  
0  Private Equity Round  20,00,00,000  
1              Series C     80,48,394

#### Setting "Sr.No" as "Index Column"

In [10]:
df.set_index('Sr No', inplace = True)

In [11]:
df.head(2)

Date dd/mm/yyyy Startup Name Industry Vertical  \
Sr No                                                  
1          09/01/2020       BYJU’S            E-Tech   
2          13/01/2020       Shuttl    Transportation   

                     SubVertical City  Location             Investors Name  \
Sr No                                                                        
1                     E-learning      Bengaluru    Tiger Global Management   
2      App based shuttle service        Gurgaon  Susquehanna Growth Equity   

            InvestmentnType Amount in USD  
Sr No                                      
1      Private Equity Round  20,00,00,000  
2                  Series C     80,48,394

#### Renaming the column name

In [20]:
df.rename(columns = {
    'Date dd/mm/yyyy': 'date',
    'Startup Name': 'startup',
    'Industry Vertical': 'vertical',
    'SubVertical': 'subvertical',
    'City  Location':  'city',
    'Investors Name':'investors',
    'InvestmentnType': 'round',
    'Amount in USD': 'amount'
}, inplace= True)

In [21]:
df.head()

date                       startup             vertical  \
Sr No                                                                  
1      09/01/2020                        BYJU’S               E-Tech   
2      13/01/2020                        Shuttl       Transportation   
3      09/01/2020                     Mamaearth           E-commerce   
4      02/01/2020  https://www.wealthbucket.in/              FinTech   
5      02/01/2020                        Fashor  Fashion and Apparel   

                                 subvertical       city  \
Sr No                                                     
1                                 E-learning  Bengaluru   
2                  App based shuttle service    Gurgaon   
3      Retailer of baby and toddler products  Bengaluru   
4                          Online Investment  New Delhi   
5                Embroiled Clothes For Women     Mumbai   

                       investors                 round        amount  
Sr No                                                                 
1        Tiger Global Management  Private Equity Round  20,00,00,000  
2      Susquehanna Growth Equity              Series C     80,48,394  
3          Sequoia Capital India              Series B   1,83,58,860  
4                 Vinod Khatumal          Pre-series A     30,00,000  
5        Sprout Venture Partners            Seed Round     18,00,000

#### Convert Amount into Crore Rupees

- As there is Missing value in "amount" only out of 3043 - 2084 rows have amount data. 
- So firstly adding zero on na place

In [26]:
df['amount'] = df['amount'].fillna('0') 

In [27]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3044 entries, 1 to 3044
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   date         3044 non-null   object
 1   startup      3044 non-null   object
 2   vertical     2873 non-null   object
 3   subvertical  2108 non-null   object
 4   city         2864 non-null   object
 5   investors    3020 non-null   object
 6   round        3040 non-null   object
 7   amount       3044 non-null   object
dtypes: object(8)
memory usage: 214.0+ KB


- Now there is no missing value in "amount column".
- Now can apply astype function to change amount type from "object to integer"

In [28]:
df['amount'].astype('int')

ValueError: invalid literal for int() with base 10: '20,00,00,000'

- We are getting "ValueError" as in amount data there is "comma" so we need to remove comma from all amount.

In [297]:
df['Startup Name'].unique()

array(['BYJU’S', 'Shuttl', 'Mamaearth', ..., 'Graphene', 'Simplotel',
       'couponmachine.in'], dtype=object)

In [31]:
df['amount'] = df['amount'].str.replace(',','')

In [32]:
df['amount']

Sr No
1       200000000
2         8048394
3        18358860
4         3000000
5         1800000
          ...    
3040      4500000
3041       825000
3042      1500000
3043            0
3044       140000
Name: amount, Length: 3044, dtype: object

- Now again trying to convert datatype into Integer after removing comma.

In [33]:
df['amount'].astype('int')


ValueError: invalid literal for int() with base 10: 'undisclosed'

- again getting error "ValueError: invalid literal for int() with base 10: 'undisclosed'" so will replace "undisclosed value with zero"

In [34]:
df['amount'] = df['amount'].str.replace('undisclosed','0')

In [35]:
df['amount'].astype('int')

ValueError: invalid literal for int() with base 10: 'unknown'

- also converting this given error "ValueError: invalid literal for int() with base 10: 'unknown'" into zero.

In [39]:
df['amount'] = df['amount'].str.replace('unknown','0')

In [40]:
df['amount'].astype('int')

ValueError: invalid literal for int() with base 10: 'Undisclosed'

- Now get error "Undisclosed" with capital "U" so will change again with "zero"

In [42]:
df['amount'] = df['amount'].str.replace('Undisclosed','0')

In [43]:
df['amount'].astype('int')

ValueError: invalid literal for int() with base 10: '4889975.54'

- Next error get "'4889975.54' so will choose amount only when its in digit only.

In [45]:
df = df[df['amount'].str.isdigit()]

In [46]:
df['amount'].astype('float')

Sr No
1       200000000.0
2         8048394.0
3        18358860.0
4         3000000.0
5         1800000.0
           ...     
3040      4500000.0
3041       825000.0
3042      1500000.0
3043            0.0
3044       140000.0
Name: amount, Length: 3029, dtype: float64

In [47]:
df['amount'] = df['amount'].astype('float')

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3029 entries, 1 to 3044
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         3029 non-null   object 
 1   startup      3029 non-null   object 
 2   vertical     2858 non-null   object 
 3   subvertical  2104 non-null   object 
 4   city         2849 non-null   object 
 5   investors    3005 non-null   object 
 6   round        3027 non-null   object 
 7   amount       3029 non-null   float64
dtypes: float64(1), object(7)
memory usage: 213.0+ KB


- Now can see 'Amount' column converted into 'float64' with '3029' rows.

In [49]:
df.head(2)

date startup        vertical                subvertical  \
Sr No                                                                  
1      09/01/2020  BYJU’S          E-Tech                 E-learning   
2      13/01/2020  Shuttl  Transportation  App based shuttle service   

            city                  investors                 round       amount  
Sr No                                                                           
1      Bengaluru    Tiger Global Management  Private Equity Round  200000000.0  
2        Gurgaon  Susquehanna Growth Equity              Series C    8048394.0

- Now after this much cange in amount finally will convert this into "crore rupees"

In [50]:
def to_inr(dollar):
    inr = dollar * 82.5
    return inr/10000000
    

In [51]:
df['amount'] =  df['amount'].apply(to_inr)

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [52]:
df['amount']

Sr No
1       1650.000000
2         66.399250
3        151.460595
4         24.750000
5         14.850000
           ...     
3040      37.125000
3041       6.806250
3042      12.375000
3043       0.000000
3044       1.155000
Name: amount, Length: 3029, dtype: float64

In [53]:
df.head(2)

date startup        vertical                subvertical  \
Sr No                                                                  
1      09/01/2020  BYJU’S          E-Tech                 E-learning   
2      13/01/2020  Shuttl  Transportation  App based shuttle service   

            city                  investors                 round      amount  
Sr No                                                                          
1      Bengaluru    Tiger Global Management  Private Equity Round  1650.00000  
2        Gurgaon  Susquehanna Growth Equity              Series C    66.39925

#### Changing date column type 

In [54]:
pd.to_datetime(df['date'])

ParserError: year 72018 is out of range: 05/072018

- While changing datetime into date type getting this error "ParserError: year 72018 is out of range: 05/072018" it visible "05/072018" slash is missing

In [55]:
df['date'] = df['date'].str.replace('05/072018', '05/07/2018')

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [57]:
df['date']

Sr No
1       09/01/2020
2       13/01/2020
3       09/01/2020
4       02/01/2020
5       02/01/2020
           ...    
3040    29/01/2015
3041    29/01/2015
3042    30/01/2015
3043    30/01/2015
3044    31/01/2015
Name: date, Length: 3029, dtype: object

In [56]:
pd.to_datetime(df['date'])

OutOfBoundsDatetime: Out of bounds nanosecond timestamp: 15-01-07 00:00:00

- after above error tried again convert datetime tyoe into date and get this error " timestamp: 15-01-07 00:00:00" 

In [58]:
pd.to_datetime(df['date'], errors = 'coerce') # "coerce will ignore the value which is nonconvertable"

Sr No
1      2020-09-01
2      2020-01-13
3      2020-09-01
4      2020-02-01
5      2020-02-01
          ...    
3040   2015-01-29
3041   2015-01-29
3042   2015-01-30
3043   2015-01-30
3044   2015-01-31
Name: date, Length: 3029, dtype: datetime64[ns]

In [59]:
df['date'] = pd.to_datetime(df['date'], errors = 'coerce')

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [60]:
df.info() # converted date in "datetime"

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3029 entries, 1 to 3044
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         3027 non-null   datetime64[ns]
 1   startup      3029 non-null   object        
 2   vertical     2858 non-null   object        
 3   subvertical  2104 non-null   object        
 4   city         2849 non-null   object        
 5   investors    3005 non-null   object        
 6   round        3027 non-null   object        
 7   amount       3029 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(6)
memory usage: 213.0+ KB


#### Removing Missing value if more than 10 percent data is miss in any column

In [63]:
df = df.dropna(subset=['date','startup','vertical','city','investors','round','amount'])

In [64]:
df.info()  # Now we have "2822 entries" after removing "missing values"

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2822 entries, 1 to 2873
Data columns (total 8 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   date         2822 non-null   datetime64[ns]
 1   startup      2822 non-null   object        
 2   vertical     2822 non-null   object        
 3   subvertical  2078 non-null   object        
 4   city         2822 non-null   object        
 5   investors    2822 non-null   object        
 6   round        2822 non-null   object        
 7   amount       2822 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(6)
memory usage: 198.4+ KB


In [65]:
df.to_csv('startup_cleaned.csv',index=False)